### ISSUE TITLE
PrimeFaces Jakarta 14.0.4 dataTable column sort does not properly work for search results

### ISSUE THREAD

**Issue by balci** [primefaces/primefaces#12675]
### Describe the bug

PrimeFaces for Jakarta EE 11 version 14.0.4 dataTable column search does not work properly for database or API search results list. 

- In listing database content in dataTable: column search works properly by clicking the headers in any sequence of actions. 

- In listing database or API search results in dataTable: column search works right after selecting a dataTable page; otherwise it does not work. Click a dataTable page and then click a column header, sort works. Click another column header, sort does not work, it does nothing.

### Reproducer

1. Go to my course assignment app: https://venus.cs.vt.edu/Books-Balci/
2. Select Favorite Books. Column sort works.
3. Select Database Search. For example, click Search Type 1, select All as category and enter king. Click Submit. Column sort does not work. Select a dataTable page. Click a column header, sort works, but thereafter it does not work unless you select another page.
4. From the menu, select Google Books API Search. In listing the API search results in dataTable, column sort behaves the same way it does for listing database search results in dataTable.

Many thanks for your excellent work! 
Prof. Osman Balci

### Expected behavior

Column sort should work regardless of where the data list is coming from in dataTable.

### PrimeFaces edition

Community

### PrimeFaces version

Jakarta 14.0.4

### JSF implementation

All

### JSF version

Jakarta 11 M3

### Java version

21.0.2

### Browser(s)

Chrome Version 129.0.6668.89

**Comment by balci** [comment]
Executable is provided at https://venus.cs.vt.edu/Books-Balci/

Here is the code that does not work:

public List<Book> getSearchItems() {

        if (searchItems == null) {
            switch (searchType) {
                case 1: // Search Type 1
                    switch (searchCategory) {
                        case "Book Title":
                            // Return the list of object references of all those books where
                            // book title contains the searchString entered by the user.
                            searchItems = bookFacade.titleQuery(searchString);
                            break;
                        case "Book Author Name":
                            // Return the list of object references of all those books where
                            // book author names contain the searchString entered by the user.
                            searchItems = bookFacade.authorQuery(searchString);
                            break;
                        case "Book Publisher Name":
                            // Return the list of object references of all those books where
                            // book publisher name contains the searchString entered by the user.
                            searchItems = bookFacade.publisherQuery(searchString);
                            break;
                        case "Book Description":
                            // Return the list of object references of all those books where
                            // book description contains the searchString entered by the user.
                            searchItems = bookFacade.descriptionQuery(searchString);
                            break;
                        default:
                            // Return the list of object references of all those books where book title OR
                            // author names OR publisher name OR description contains the searchString entered by the user.
                            searchItems = bookFacade.allQuery(searchString);
                    }
                    break;
                case 2: // Search Type 2
                    // Book publication date ≥ dateQ
                    java.sql.Date sqlDateQ2 = new java.sql.Date(dateQ.getTime());
                    searchItems = bookFacade.type2SearchQuery(sqlDateQ2);
                    break;
                case 3: // Search Type 3
                    // Book publication date ≤ dateQ
                    java.sql.Date sqlDateQ3 = new java.sql.Date(dateQ.getTime());
                    searchItems = bookFacade.type3SearchQuery(sqlDateQ3);
                    break;
                case 4: // Search Type 4
                    // Book publication date is between minDateQ and maxDateQ inclusive
                    java.sql.Date sqlMinDateQ = new java.sql.Date(minDateQ.getTime());
                    java.sql.Date sqlMaxDateQ = new java.sql.Date(maxDateQ.getTime());
                    searchItems = bookFacade.type4SearchQuery(sqlMinDateQ, sqlMaxDateQ);
                    break;
                case 5: // Search Type 5
                    // Book page count ≥ pageCountQ
                    searchItems = bookFacade.type5SearchQuery(pageCountQ);
                    break;
                case 6: // Search Type 6
                    // Book page count ≤ pageCountQ
                    searchItems = bookFacade.type6SearchQuery(pageCountQ);
                    break;
                case 7: // Search Type 7
                    // Book page count is between minPageCountQ and maxPageCountQ inclusive
                    searchItems = bookFacade.type7SearchQuery(minPageCountQ, maxPageCountQ);
                    break;
                default:
                    Methods.showMessage("Fatal Error", "Search Type is Out of Range!",
                            "");
            }
        }

        // Sort the List of searchItems objects with respect to book title property in alphabetical order
        searchItems.sort(Comparator.comparing(Book::getTitle));

        return searchItems;
    }

**Comment by melloware** [comment]
i think its less about how you are querying and how you are storing that `List` for the Datatable. Scope etc. Without a reproducer we would be just guess but i am pretty sure its an issue with your code.

**Comment by FlipWarthog** [comment]
@balci 

If you move
```java
// Sort the List of searchItems objects with respect to book title property in alphabetical order
searchItems.sort(Comparator.comparing(Book::getTitle));
```

to be within the `if (searchItems == null)` block, does it work?

**Comment by balci** [comment]
Moving the sort statement within the if (searchItems == null) block solves the problem. Now the dataTable column sort works very nicely! THANKS!!!

**Comment by balci** [comment]
On behalf of 81 students and myself, we thank you for resolving this issue.

Osman Balci
Director of Mobile/Cloud Software Engineering Lab
Professor of Computer Science
https://manta.cs.vt.edu/balci
Tel: (540) 231-4841    ***@***.******@***.***>
638 McBryde Hall, MC 0106
Department of Computer Science
Virginia Tech, Blacksburg, VA 24061, U.S.A.


From: Melloware ***@***.***>
Date: Thursday, October 10, 2024 at 2:34 PM
To: primefaces/primefaces ***@***.***>
Cc: Balci, Osman ***@***.***>, Mention ***@***.***>
Subject: Re: [primefaces/primefaces] PrimeFaces Jakarta 14.0.4 dataTable column sort does not properly work for search results (Issue #12675)

Closed #12675<https://github.com/primefaces/primefaces/issues/12675> as not planned.
